swift/ql/lib/codeql/swift/security/XXE.qll

```diff
@@ -90,3 +91,73 @@ private class NodeLoadExternalEntitiesAlways extends VarDecl {
     this.getEnclosingDecl().(StructDecl).getFullName() = "XMLNode.Options"
   }
 }
+
+/** The XML argument of an `AEXMLDocument` vulnerable to XXE. */
+private class AexmlDocumentSink extends XxeSink {
+  AexmlDocumentSink() {
+    // `AEXMLDocument` initialized with vulnerable options.
+    exists(ApplyExpr call | this.asExpr() = call.getArgument(0).getExpr() |
+      call.(VulnerableAexmlDocument)
+          .getStaticTarget()
+          .hasName(["init(xml:options:)", "init(xml:encoding:options:)"])
+      or
+      // `loadXML` called on a vulnerable AEXMLDocument.
+      DataFlow::localExprFlow(any(VulnerableAexmlDocument v),
+        call.(AexmlDocumentLoadXml).getQualifier())
+    )
+  }
+}
+
+/** The XML argument of an `AEXMLParser` initialized with an `AEXMLDocument` vulnerable to XXE. */
+private class AexmlParserSink extends XxeSink {
+  AexmlParserSink() {
+    exists(AexmlParser parser | this.asExpr() = parser.getArgument(1).getExpr() |
+      DataFlow::localExprFlow(any(VulnerableAexmlDocument v), parser.getArgument(0).getExpr())
+    )
+  }
+}
+
+/** The creation of an `AEXMLDocument` that receives a vulnerable `AEXMLOptions` argument. */
+private class VulnerableAexmlDocument extends AexmlDocument {
+  VulnerableAexmlDocument() {
+    exists(AexmlOptions optionsArgument, VulnerableAexmlOptions vulnOpts |
+      this.getAnArgument().getExpr() = optionsArgument and
+      DataFlow::localExprFlow(vulnOpts, optionsArgument)
+    )
+  }
+}
+
+/**
+ * An `AEXMLOptions` object which contains a `parserSettings` with `shouldResolveExternalEntities`
+ * set to `true`.
+ */
+private class VulnerableAexmlOptions extends AexmlOptions {
+  VulnerableAexmlOptions() {
+    exists(ParserSettings parserSettings, AexmlShouldResolveExternalEntities sree, AssignExpr a |
+      a.getSource() = any(BooleanLiteralExpr b | b.getValue() = true) and
+      a.getDest() = sree and
+      sree.(MemberRefExpr).getBase() = parserSettings and
+      parserSettings.(MemberRefExpr).getBase() = this
+    )
+  }
+}
+
+/** An expression of type `AEXMLOptions.ParserSettings`. */
+class ParserSettings extends Expr {
+  pragma[inline]
```

Comment:
These two `pragma[inlines]` are an attempt at avoiding fully computing these classes, which seemed to be potentially too big depending on the codebase. Hopefully inlining them where they are used should add the appropriate context to prevent it, while allowing us to reuse this code, but please let me know if this solution is not desirable performance-wise and/or there's a better alternative. I'll launch DCA in any case to see how this behaves.

Reply:
Yep, I agree we should do a DCA run to check how this does.  Though I'm not _overly_ concerned about Swift query performance - I think it will be more effective to do a review later on (when we have a full query suite and thus better data).